platforms/android/library/src/main/java/io/element/android/wysiwyg/utils/HtmlToSpansParser.kt

```diff
@@ -46,464 +42,227 @@ internal class HtmlToSpansParser(
-     * `li` one will be parsed and its span will be added. However, as they were added in reversed
-     * order (`quote > li` instead of `li > quote`), it will appear as a list item inside a quote
-     * when Android renders the resulting text.
-     *
-     * To fix that, we're creating this list of Spans to be added to the text:
-     * 1. When we parse the start tag, we set a placeholder span to 'book' the position.
-     * 2. When we parse the end tag, we replace that placeholder with the real span, keeping the
-     * starting position.
-     * 3. Once we've parsed the whole HTML, we apply these spans to the text in order.
-     *
-     * *Note*: this is only needed for block spans, as inline spans can be rendered in any order.
+     * Convert the HTML string into a [Spanned] text.
      */
-    private val spansToAdd = mutableListOf<PendingSpan<*>>()
-
-    private val parser = Parser().also { it.contentHandler = this }
-    private val text = SpannableStringBuilder()
-
     fun convert(): Spanned {
-        spansToAdd.clear()
-        parser.parse(InputSource(StringReader(html)))
-        if (text.isNotEmpty()) {
-            for (spanToAdd in spansToAdd) {
-                val start = spanToAdd.start.coerceIn(0, text.length)
-                val end = spanToAdd.end.coerceIn(0, text.length)
-                text.setSpan(spanToAdd.span, start, end, spanToAdd.flags)
-            }
-            text.removePlaceholderSpans()
-            text.addAtRoomSpans()
+        val dom = Jsoup.parse(html)
+        val text = buildSpannedString {
+            val body = dom.body()
+            parseChildren(body)
+            addAtRoomSpans()
         }
         if (BuildConfig.DEBUG) text.assertOnlyAllowedSpans()
         return text
     }
 
-    override fun setDocumentLocator(locator: Locator?) {}
-
-    override fun startDocument() {}
+    private fun SpannableStringBuilder.parseChildren(element: Element) {
+        for (child in element.childNodes()) {
+            when (child) {
+                is Element -> parseElement(child)
+                is TextNode -> parseTextNode(child)
+            }
+        }
+    }
 
-    override fun endDocument() {}
+    private fun SpannableStringBuilder.parseElement(element: Element) {
+        when (element.tagName()) {
+            "a" -> parseLink(element)
+            "b", "strong" -> parseInlineFormatting(element, InlineFormat.Bold)
+            "i", "em" -> parseInlineFormatting(element, InlineFormat.Italic)
+            "u" -> parseInlineFormatting(element, InlineFormat.Underline)
+            "del" -> parseInlineFormatting(element, InlineFormat.StrikeThrough)
+            "code" -> parseInlineCode(element)
+            "ul", "ol" -> parseChildren(element)
+            "li" -> parseListItem(element)
+            "pre" -> parseCodeBlock(element)
+            "blockquote" -> parseQuote(element)
+            "p" -> parseParagraph(element)
+            "br" -> parseLineBreak(element)
```

Comment:
Maybe add a log for un-handled tag names?

Reply:
Good call! In fact, I added a clean step where we remove any tags and attributes we don't recognise.